File: shared/ios/KeybaseShare/ShareViewController.m
//
//  ShareViewController.m
//  KeybaseShare
//
//  Created by Michael Maxim on 8/31/18.
//  Copyright © 2018 Keybase. All rights reserved.
//

#import "ShareViewController.h"
#import "keybase/keybase.h"
#import <MobileCoreServices/MobileCoreServices.h>
#import "ItemProviderHelper.h"

#if TARGET_OS_SIMULATOR
const BOOL isSimulator = YES;
#else
const BOOL isSimulator = NO;
#endif


@interface ShareViewController ()
@property (nonatomic, strong) ItemProviderHelper * iph;
@property UIAlertController* alert;
@end

@implementation ShareViewController

- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
}

- (void) openApp {
  NSURL *url = [NSURL URLWithString:@"keybase://incoming-share"];
  UIResponder *responder = self;
  while (responder) {
        if ([responder respondsToSelector: @selector(openURL:)]){
          @try {
            // This is needed by ios18+ to function now on the new xcode. This FAILs in the simulator but seems
            // to work on device
            NSMethodSignature *signature = [responder methodSignatureForSelector:@selector(openURL:options:completionHandler:)];
            NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
            [invocation setTarget:responder];
            [invocation setSelector:@selector(openURL:options:completionHandler:)];
            
            NSDictionary *options = @{};
            void (^completionHandler)(BOOL success) = nil;
            
            [invocation setArgument:&url atIndex:2];
            [invocation setArgument:&options atIndex:3];
            [invocation setArgument:&completionHandler atIndex:4];
            [invocation invoke];
            return;
          }
          @catch (NSException *exception) {
            NSLog(@"Exception occurred while opening URL to share");
          }
        }
      responder = [responder nextResponder];
  }
}

- (void) completeRequestAlreadyInMainThread {
  [self.alert dismissViewControllerAnimated:true completion:^{
    [self.extensionContext completeRequestReturningItems:nil completionHandler:^(BOOL expired) {
      [self openApp];
    }];
  }];
}

- (void)showProgressView {
  UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Working on it"
                                                                        message:@"\n\nPreparing content for sharing into Keybase."
                                                                 preferredStyle:UIAlertControllerStyleAlert];

  self.alert = alertController;
  UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
  [spinner startAnimating];
  spinner.translatesAutoresizingMaskIntoConstraints = NO;

  [alertController.view addSubview:spinner];

  [spinner.centerXAnchor constraintEqualToAnchor:alertController.view.centerXAnchor].active = YES;
  [spinner.centerYAnchor constraintEqualToAnchor:alertController.view.centerYAnchor constant:-8].active = YES;

  [self presentViewController:alertController animated:YES completion:nil];
}

- (void) closeProgressView {
  [self.alert dismissViewControllerAnimated:true completion:nil];
}

- (void) viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];
  
  NSMutableArray *itemArrs = [NSMutableArray array];
  for (NSExtensionItem *inputItem in self.extensionContext.inputItems) {
    [itemArrs addObject:inputItem.attachments];
  }
  
  __weak __typeof__(self) weakSelf = self;
  self.iph = [[ItemProviderHelper alloc] initForShare: true withItems:itemArrs completionHandler:^{
    __typeof__(self) strongSelf = weakSelf;
    if (strongSelf != nil) {
      [strongSelf completeRequestAlreadyInMainThread];
    }
  }];
  [self showProgressView];
  [self.iph startProcessing];
}

@end

File: shared/ios/KeybaseShare/ShareViewController.swift
//
//  ShareViewController.swift
//  KeybaseShare
//
//  Created by Chris Nojima on 6/12/25.
//  Copyright © 2025 Keybase. All rights reserved.
//

import Foundation
import Intents
import UIKit
import MobileCoreServices
import Keybasego
import KBCommon

@objc(ShareViewController)
public class ShareViewController: UIViewController {
  var iph: ItemProviderHelper?
  var alert: UIAlertController?
  var selectedConvID: String?

  public override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }

  func openApp() {
    let path = selectedConvID.map { "keybase://incoming-share/\($0)" } ?? "keybase://incoming-share"
    guard let url = URL(string: path) else { return }
    var responder: UIResponder? = self
    while let r = responder {
      if r.responds(to: #selector(UIApplication.openURL(_:))) {
        do {
          let sel = #selector(UIApplication.open(_:options:completionHandler:))
          if r.responds(to: sel) {
            let imp = r.method(for: sel)
            typealias Func = @convention(c) (AnyObject, Selector, URL, [UIApplication.OpenExternalURLOptionsKey: Any], ((Bool) -> Void)?) -> Void
            let f = unsafeBitCast(imp, to: Func.self)
            f(r, sel, url, [:], nil)
            return
          }
        }
      }
      responder = r.next
    }
  }

  func completeRequestAlreadyInMainThread() {
    alert?.dismiss(animated: true) {
      DispatchQueue.main.async {
        self.openApp()
        self.extensionContext?.completeRequest(returningItems: nil, completionHandler: nil)
      }
    }
  }

  func showProgressView() {
    let alertController = UIAlertController(title: "Working on it", message: "\n\nPreparing content for sharing into Keybase.", preferredStyle: .alert)
    alert = alertController
    let spinner = UIActivityIndicatorView(style: .medium)
    spinner.translatesAutoresizingMaskIntoConstraints = false
    spinner.startAnimating()
    alertController.view.addSubview(spinner)
    NSLayoutConstraint.activate([
      spinner.centerXAnchor.constraint(equalTo: alertController.view.centerXAnchor),
      spinner.centerYAnchor.constraint(equalTo: alertController.view.centerYAnchor, constant: -8)
    ])
    present(alertController, animated: true, completion: nil)
  }

  func closeProgressView() {
    alert?.dismiss(animated: true, completion: nil)
  }

  public override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    if let intent = extensionContext?.intent as? INSendMessageIntent {
      selectedConvID = intent.conversationIdentifier
    }
    let itemArrs = extensionContext?.inputItems.compactMap {
      ($0 as? NSExtensionItem)?.attachments
    } ?? []

    weak var weakSelf = self
    iph = ItemProviderHelper(forShare: true, withItems: itemArrs) {
      guard let self = weakSelf else { return }
      self.completeRequestAlreadyInMainThread()
    }
    showProgressView()
    iph?.startProcessing()
  }
}
